contracts/LivepeerProtocol.sol

```diff
@@ -523,11 +540,87 @@ contract LivepeerProtocol {
     }
 
     /*
-     * Pseudorandomly elect an active transcoder. Currently a placeholder
-     * TODO: take into account pricing, etc.
+     * Submit a transcoding job
+     * @param _streamId Unique stream identifier
+     * @param _transcodingOptions Output bitrates, formats, encodings
+     * @param _maxPricePerSegment Max price (in LPT base units) to pay for transcoding a segment of a stream
+     */
+    function job(uint256 _streamId, bytes32 _transcodingOptions, uint256 _maxPricePerSegment) returns (bool) {
+        address electedTranscoder = electCurrentActiveTranscoder(_maxPricePerSegment);
+
+        // Check if there is an elected current active transcoder
+        if (electedTranscoder == address(0)) throw;
+
+        jobs.push(Job(jobs.length - 1, _streamId, _transcodingOptions, _maxPricePerSegment, msg.sender, electedTranscoder, 0));
+
+        return true;
+    }
+
+    /*
+     * Computes the status of a job
+     * @param _jobId Job identifier
+     */
+    function jobStatus(uint256 _jobId) constant returns (JobStatus) {
+        if (jobs[_jobId].endBlock > 0 && jobs[_jobId].endBlock <= block.number) {
+            // A job is inactive if its end block is set and the current block is greater than or equal to the job's end block
+            return JobStatus.Inactive;
+        } else {
+            // A job is active if the current block is less than the job's termination block
+            return JobStatus.Active;
+        }
+    }
+
+    /*
+     * Return a job
+     * @param _jobId Job identifier
+     */
+    function getJob(uint256 _jobId) constant returns (uint256, uint256, bytes32, uint256, address, address, uint256) {
+        Job job = jobs[_jobId];
+
+        return (job.jobId, job.streamId, job.transcodingOptions, job.maxPricePerSegment, job.broadcasterAddress, job.transcoderAddress, job.endBlock);
+    }
+
+    /*
+     * End a job. Can be called by either a broadcaster or transcoder of a job
+     */
+    function endJob(uint256 _jobId) returns (bool) {
+        // Check if job exists
+        if (_jobId < 0 || _jobId >= jobs.length) throw;
+        // Check if job already has an end block
+        if (jobs[_jobId].endBlock > 0) throw;
+        // Check if called by either broadcaster or transcoder
+        if (msg.sender != jobs[_jobId].broadcasterAddress && msg.sender != jobs[_jobId].transcoderAddress) throw;
+
+        // Set set end block for job to be in 10 blocks
+        // Note: this is a placeholder value and subject to change
+        jobs[_jobId].endBlock = block.number + 10;
```

Comment:
There could be a scenario in which a broadcaster has streamed all segments to a transcoder and then calls `endJob()`, but the transcoder has not had a chance to submit transcode claims for all the segments yet. The `endBlock` field for a job gives a transcoder a grace period during which it can finish submitting transcode claims.